eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Proposals.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Test.Cardano.Ledger.Conway.Proposals where
+
+import Cardano.Ledger.Conway
+import Cardano.Ledger.Conway.Governance
+import Control.DeepSeq (force)
+import Control.Exception (evaluate)
+import Data.Foldable (foldl', toList)
+import qualified Data.Map.Strict as Map
+import Data.Maybe (fromMaybe)
+import Data.Sequence (fromList)
+import qualified Data.Sequence as Seq
+import Data.Sequence.Strict (StrictSeq (Empty, (:<|), (:|>)))
+import qualified Data.Set as Set
+import Test.Cardano.Ledger.Common
+import Test.Cardano.Ledger.Conway.Arbitrary (
+  ProposalsForEnactment (..),
+  ProposalsNewActions (..),
+ )
+
+spec :: Spec
+spec = do
+  describe "Proposals" $ do
+    context "Construction" $ do
+      prop "Generator generates consistent Proposals" $
+        \(ps :: Proposals Conway) ->
+          ps `shouldSatisfy` proposalsAreConsistent
+      prop "Adding new nodes keeps Proposals consistent" $
+        \(ProposalsNewActions ps actions :: ProposalsNewActions Conway) ->
+          let ps' =
+                foldl'
+                  (\p action -> fromMaybe (error "Unable to add action") $ proposalsAddAction action p)
+                  ps
+                  actions
+           in ps' `shouldSatisfy` proposalsAreConsistent
+    context "Removal" $ do
+      prop "Removing leaf nodes keeps Proposals consistent" $
+        \(ps :: Proposals Conway) ->
+          let test gais =
+                let ps' = fst $ proposalsRemoveDescendentIds gais ps
+                 in do
+                      proposalsSize ps' `shouldBe` proposalsSize ps - Set.size gais
+                      ps' `shouldSatisfy` proposalsAreConsistent
+           in case proposalsIds ps of
+                Empty -> True `shouldBe` True
+                _pids :|> pid :|> pid1 :|> pid2 :|> pid3 -> test $ Set.fromList [pid, pid1, pid2, pid3]
```

Comment:
I think it'd help with readability if there were functions like `takeInit` and `takeTail` that take the first n elements from either the beginning or the end of the seq.